tdi_python/tdiTable.py

```diff
@@ -0,0 +1,2617 @@
+
+
+class TdiTable:
+
+    """
+    This class manages the exchange of information between
+    the CLI and BF Runtime's C API. It has two primary parts:
+      - Looking up the action and field metadata for a table
+      - Parsing and reformatting data as the user makes calls
+
+    Note that keys in this object are the c-string representation
+    (byte-streams in python) of data, not python strings.
+    """
+    def __init__(self, cintf, handle, info):
+        self._cintf = cintf
+        self._tdi_info = info
+        self._handle = handle
+        # get get_device
+        self._device_hdl = cintf.get_device()
+        self.key_field_readables = []
+        self.data_field_readables = []
+        self.action_readables = {}
+        self.action_data_readables = {}
+        self.tdi_id = -1
+        self.frontend = None
+        self.idle_tmo_callback = None
+        self.register_sync_cb = None
+        self.counter_sync_cb = None
+        self.hit_state_update_cb = None
+        self.port_status_notif_callback = None
+        self.selector_table_update_cb = None
+        self.string_choices = {}
+        self.annotations = {}
+        self.has_const_default_action = False
+        self.supported_commands = ["info", "add_from_json", "entry", "string_choices"]
+        #self.set_supported_attributes_to_supported_commands()
+        self.set_supported_operations_to_supported_commands()
+        self.set_supported_apis_to_supported_commands()
+        self.unimplemented_tables = ["TODO"]
+        self.table_type = self.table_type_map((self.get_type()))
+        self.table_ready = False if self.table_type in self.unimplemented_tables else True
+        #
+        # Ignore INVALID Table Type from Driver
+        #
+        if self.table_type_map(self.get_type()) == "INVLD":
+            return
+        #
+        # Determine the **Table Name** from Driver
+        #
+        table_name = c_char_p()
+        sts = self._cintf.get_driver().tdi_table_name_get(self._handle, byref(table_name))
+        #pdb.set_trace()
+        if not sts == 0:
+            print("CLI Error: get table name failed. [{}]".format(self._cintf.err_str(sts)))
+            raise TdiTableError("Table init name failed.", None, -1)
+        self.name = table_name.value.decode('ascii')
+        #Unify the table name for BFNode (command nodes)
+        if self.table_type in ["PORT_CFG", "PORT_STAT", "PORT_HDL_INFO", "PORT_FRONT_PANEL_IDX_INFO", "PORT_STR_INFO"]:
+            self.name = "port.{}".format(self.name.lower().replace("$",""))
+        if self.table_type in ["PRE_MGID", "PRE_NODE", "PRE_ECMP", "PRE_LAG", "PRE_PRUNE", "PRE_PORT", "MIRROR_CFG"]:
+            self.name = self.name.lower().replace("$","")
+        if self.table_type in ["TM_PORT_GROUP_CFG", "TM_PORT_GROUP"]:
+            self.name = self.name.lower().replace("$","")
+        if self.table_type in ["SNAPSHOT", "SNAPSHOT_LIVENESS"]:
+            self.name = "{}".format(self.name.lower().replace("$",""))
```

Comment:
Can we please remove all these table specific code? Maybe check the table name, if it is caps and $, convert it?

Reply:
Since it is based on the port type to decide whether need prefix or not for table name.
The name conversion logic is the same: so just convert the name at the begin.
name_lowercase_without_dollar=self.name.lower().replace("$","")